lib/Dialect/FIRRTL/NLATable.cpp

```diff
@@ -0,0 +1,116 @@
+
+//===- NLATable.cpp - Non-Local Anchor Table --------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/NLATable.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "mlir/IR/BuiltinOps.h"
+
+using namespace circt;
+using namespace firrtl;
+
+NLATable::NLATable(Operation *operation) {
+  if (auto mod = dyn_cast<mlir::ModuleOp>(operation))
+    for (auto &op : *mod.getBody())
+      if ((operation = dyn_cast<CircuitOp>(&op)))
+        break;
+
+  auto circuit = cast<CircuitOp>(operation);
+  // We are assuming it's faster to iterate over the top level twice than cache
+  // a large number of options.
+  for (auto &op : *circuit.getBody()) {
+    if (auto module = dyn_cast<FModuleLike>(op))
+      symToOp[module.moduleNameAttr()] = module;
+    if (auto nla = dyn_cast<NonLocalAnchor>(op)) {
+      symToOp[nla.sym_nameAttr()] = nla;
+      for (auto ent : nla.namepath()) {
+        if (auto mod = ent.dyn_cast<FlatSymbolRefAttr>())
+          nodeMap[mod.getAttr()].push_back(nla);
+        else if (auto inr = ent.dyn_cast<hw::InnerRefAttr>())
+          nodeMap[inr.getModule()].push_back(nla);
+      }
+    }
+  }
+}
+
+// void NLATable::replaceInstance(InstanceOp oldInst, InstanceOp newInst) {
+//     assert(oldInst.moduleName() == newInst.moduleName() && "module targets
+//     mismatch"); assert(oldInst->getParentOfType<FModuleOp>() ==
+//     newInst->getParentOfType<FModuleOp>() && "module targets mismatch"); auto
+//     nlas = nodeMap.equal_range(oldInst->getParentOfType<FModuleOp>()); for
+//     (auto i = nlas.first, e = nlas.second; i != e; ++i)
+//         i->second.updateInstance(oldInst, newInst);
+// }
```

Comment:
Should this be part of the public API?

Reply:
Yes, but I don't have a user of it yet, so deleted.  Didn't mean to leave it in :)